### PR TITLE
Kube-Proxy: Track packets accepted on localhost nodeports

### PR DESCRIPTION
**Pull request by aroradaman**
#### What type of PR is this?
Part of nftables-todo https://github.com/kubernetes/kubernetes/issues/122572
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
The iptables mode of kube-proxy now tracks accepted packets that are destined for node-ports on localhost by introducing `kubeproxy_iptables_localhost_nodeports_accepted_packets_total` metric.
This will help users to identify if they rely on iptables.localhostNodePorts feature and ulitmately help them to migrate from iptables to nftables.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig network
/assign @danwinship @aojea 

```
root@kind-worker:/# nfacct list
{ pkts = 00000000000000000000, bytes = 00000000000000000000 } = ct_state_invalid_dropped_pkts;
{ pkts = 00000000000000000000, bytes = 00000000000000000000 } = localhost_nps_accepted_pkts;

root@kind-worker:/# for i in {1..4321}; do curl --silent localhost:31490 > /dev/null; done;

root@kind-worker:/# nfacct list
{ pkts = 00000000000000000000, bytes = 00000000000000000000 } = ct_state_invalid_dropped_pkts;
{ pkts = 00000000000000004321, bytes = 00000000000000259260 } = localhost_nps_accepted_pkts;

root@kind-worker:/# curl --silent 127.0.0.1:10249/metrics | grep kubeproxy_iptables_localhost_nodeports_accepted_packets_total
# HELP kubeproxy_iptables_localhost_nodeports_accepted_packets_total [ALPHA] Number of packets accepted on nodeports of loopback interface
# TYPE kubeproxy_iptables_localhost_nodeports_accepted_packets_total counter
kubeproxy_iptables_localhost_nodeports_accepted_packets_total 4321

root@kind-worker:/# 
```

### CODE REVIEW

File: pkg/proxy/metrics/metrics.go [hunk 1]
```diff
@@ -289,7 +289,7 @@ func RegisterMetrics(mode kubeproxyconfig.ProxyMode) {
 
 		switch mode {
 		case kubeproxyconfig.ProxyModeIPTables:
-			legacyregistry.CustomMustRegister(newCTStateInvalidPacketsCollector())
+			legacyregistry.CustomMustRegister(newNFAcctMetricCollector(IPTablesCTStateInvalidDroppedNFAcctCounter, iptablesCTStateInvalidDroppedPacketsDescription))
```
Comment (danwinship) [inline pkg/proxy/metrics/metrics.go:300]: could maybe have a top-level var

    iptablesCTStateInvalidDroppedPackets = newNFAcctMetricCollector(IPTablesCTStateInvalidDroppedNFAcctCounter, iptablesCTStateInvalidDroppedPacketsDescription)

so you can just do `legacyregistry.CustomMustRegister(iptablesCTStateInvalidDroppedPackets)` here?

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -1183,6 +1186,22 @@ func (proxier *Proxier) syncProxyRules() {
...
 				// and we can't change that.
+				if proxier.localhostNodePorts && proxier.ipFamily == v1.IPv4Protocol {
+					rule := []string{
+						"-A", string(kubeNodePortsChain),
+						"-m", "comment", "--comment", svcPortNameString,
+						"-m", protocol, "-p", protocol,
+						"-d", "127.0.0.0/8",
+						"--dport", strconv.Itoa(svcInfo.NodePort()),
+					}
+					if proxier.nfAcctCounters[metrics.LocalhostNodePortAcceptedNFAcctCounter] {
```
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go:1197]: you should merge this check in with the `if` above; we only need a separate 127.0.0.0/8 rule if the nfacct counter exists

**Review by danwinship — COMMENTED** [review]
Looks good, but why does this not break unit tests? Oh, I guess because the nfacct counters don't get created in that case? Maybe `NewFakeProxier` should fake that, so we get the ruleset with the nfacct rules?

Also, weren't you going to bring back the ctstate nfacct e2e test to test this case since it shouldn't run into the problem with calico that the ctstate one did?

File: pkg/proxy/metrics/metrics.go (hunk 1 above)
Comment (aroradaman) [reply pkg/proxy/metrics/metrics.go:300]: sure

Comment (aroradaman) [reply pkg/proxy/metrics/metrics.go:300]: Should I define newNFAcctMetricCollector(...) where the metric description and counter name is defined instead? 
That will group all the metric-related objects.

```go
	// localhostNodePortsAcceptedPacketsDescription describe the metrics for the number of packets accepted
	// by iptables which were destined for nodeports on loopback interface.
	localhostNodePortsAcceptedPacketsDescription = metrics.NewDesc(
		"kubeproxy_iptables_localhost_nodeports_accepted_packets_total",
		"Number of packets accepted on nodeports of loopback interface",
		nil, nil, metrics.ALPHA, "")
	LocalhostNodePortAcceptedNFAcctCounter = "localhost_nps_accepted_pkts"
	localhostNodePortAcceptedMetricCollector = newNFAcctMetricCollector(...)
	
	
	
	
	legacyregistry.CustomMustRegister(localhostNodePortAcceptedMetricCollector)
	
```

**Comment by aroradaman** [comment]
In `pull-kubernetes-e2e-kind-nftables`  **iptables** is returned as proxyMode. Looking into this.
```
I0521 17:58:56.655573 66562 builder.go:121] Running '/home/prow/go/src/k8s.io/kubernetes/_output/bin/kubectl --server=https://127.0.0.1:36149/ --kubeconfig=/root/.kube/kind-test-config --namespace=kube-proxy-6402 exec host-exec-pod -- /bin/sh -x -c curl --silent 127.0.0.1:10249/proxyMode'
I0521 17:58:56.944047 66562 builder.go:146] stderr: "+ curl --silent 127.0.0.1:10249/proxyMode\n"
I0521 17:58:56.944093 66562 builder.go:147] stdout: "iptables"
```

**Comment by aojea** [comment]
> 2024-05-22T12:39:07.236239492Z stderr F I0522 12:39:07.234678       1 server_linux.go:292] "Using nftables Proxier"

now is working ok, thanks for catching this

**Comment by danwinship** [comment]
wait, the nftables job was running iptables? :man_facepalming:

**Comment by aojea** [comment]
> wait, the nftables job was running iptables? 🤦‍♂️

when we switched to beta, I disable the experimental job and forget to add the env variables to this job 😅  it ran nftables before that

**Comment by dashpole** [comment]
/assign
(for sig-instrumentation review)
/triage accepted

File: pkg/proxy/metrics/metrics.go [hunk 3]
```diff
@@ -265,6 +266,15 @@ var (
 		},
 		[]string{"traffic_policy"},
 	)
+
+	// localhostNodePortsAcceptedPacketsDescription describe the metrics for the number of packets accepted
+	// by iptables which were destined for nodeports on loopback interface.
+	localhostNodePortsAcceptedPacketsDescription = metrics.NewDesc(
+		"kubeproxy_iptables_localhost_nodeports_accepted_packets_total",
+		"Number of packets accepted on nodeports of loopback interface",
+		nil, nil, metrics.ALPHA, "")
```
Comment (dashpole) [inline pkg/proxy/metrics/metrics.go:275]: Just confirming: ALPHA means you intend to graduate this metric to BETA and then to GA in the next few releases.  If this is one you are using to develop or debug, and don't intend to promote, use INTERNAL instead.

**Review by dashpole — APPROVED** [review]

File: pkg/proxy/metrics/metrics.go (hunk 3 above)
Comment (danwinship) [reply pkg/proxy/metrics/metrics.go:275]: It's not internal; this is for user to use (to figure out if they are depending on the localhost nodeports functionality that iptables kube-proxy implements but nftables kube-proxy does not).

As for graduation; _all_ of the kube-proxy metrics are alpha and there has never been any discussion about graduating any of them :grimacing:

**Comment by saschagrunert** [comment]
@aroradaman we're hitting an issue described in https://github.com/cri-o/cri-o/issues/8270#issuecomment-2157613376

The metric should be registered in our test environment but kube proxy fails to sync the proxy rules. Do you have any further idea what could be the root cause? An specific kernel or iptables version?

**Pull request merged**